src/webgpu/api/validation/capability_checks/limits/limit_utils.ts

```diff
@@ -45,6 +39,194 @@ const EncoderTypes = {
+  order: Order,
+  bindGroupTest: BindGroupTest,
+  storageDefinitionWGSLSnippetFn: (i: number, j: number) => string,
+  bodyFn: (numBindings: number, set: number) => string,
+  numBindings: number,
+  extraWGSL = ''
+) {
+  switch (bindingCombination) {
+    case 'vertex':
+      return `
+        ${extraWGSL}
+        ${getWGSLBindings(order, bindGroupTest, storageDefinitionWGSLSnippetFn, numBindings, 0)}
+        @vertex fn mainVS() -> @builtin(position) vec4f {
+          ${bodyFn(numBindings, 0)}
+        }
+      `;
+    case 'fragment':
+      return `
+        ${extraWGSL}
+        ${getWGSLBindings(order, bindGroupTest, storageDefinitionWGSLSnippetFn, numBindings, 0)}
+        @vertex fn mainVS() -> @builtin(position) vec4f {
+          return vec4f(0);
+        }
+        @fragment fn mainFS() -> @location(0) vec4f {
+          ${bodyFn(numBindings, 0)}
+        }
+      `;
+    case 'vertexAndFragmentWithPossibleVertexStageOverflow': {
+      return `
+        ${extraWGSL}
+        ${getWGSLBindings(order, bindGroupTest, storageDefinitionWGSLSnippetFn, numBindings, 0)}
+        ${getWGSLBindings(order, bindGroupTest, storageDefinitionWGSLSnippetFn, numBindings - 1, 1)}
+        @vertex fn mainVS() -> @builtin(position) vec4f {
+          ${bodyFn(numBindings, 0)}
+        }
+        @fragment fn mainFS() -> @location(0) vec4f {
+          ${bodyFn(numBindings - 1, 1)}
+        }
+      `;
+    }
+    case 'vertexAndFragmentWithPossibleFragmentStageOverflow': {
+      return `
+        ${extraWGSL}
+        ${getWGSLBindings(order, bindGroupTest, storageDefinitionWGSLSnippetFn, numBindings - 1, 0)}
+        ${getWGSLBindings(order, bindGroupTest, storageDefinitionWGSLSnippetFn, numBindings, 1)}
+        @vertex fn mainVS() -> @builtin(position) vec4f {
+          ${bodyFn(numBindings - 1, 0)}
+        }
+        @fragment fn mainFS() -> @location(0) vec4f {
+          ${bodyFn(numBindings, 1)}
+        }
+      `;
+    }
+    case 'compute':
+      return `
+        ${extraWGSL}
+        ${getWGSLBindings(order, bindGroupTest, storageDefinitionWGSLSnippetFn, numBindings, 0)}
+        @group(3) @binding(0) var<storage, read_write> d: f32;
+        @compute @workgroup_size(1) fn main() {
+          ${bodyFn(numBindings, 0)}
+        }
+      `;
+      break;
+  }
+}
+
+export function getPerStageWGSLForBindingCombination(
+  bindingCombination: BindingCombination,
+  order: Order,
+  bindGroupTest: BindGroupTest,
+  storageDefinitionWGSLSnippetFn: (i: number, j: number) => string,
+  usageWGSLSnippetFn: (i: number, j: number) => string,
+  numBindings: number,
+  extraWGSL = ''
+) {
+  return getPerStageWGSLForBindingCombinationImpl(
+    bindingCombination,
+    order,
+    bindGroupTest,
+    storageDefinitionWGSLSnippetFn,
+    (numBindings: number, set: number) => {
+      return bindingCombination === 'compute'
+        ? `d = ${range(numBindings, i => usageWGSLSnippetFn(i, set)).join(' +\n')};`
```

Comment:
Another possibility that could be simpler is to use a bunch of phony assignments

Reply:
thanks. I didn't know about `_` until recently 😅

fixed.